cpp/src/arrow/compute/kernels/scalar_arithmetic.cc

```diff
@@ -852,24 +853,243 @@ struct LogbChecked {
+    return std::round(val * T(0.5)) * 2;
+  }
+};
+
+template <typename T>
+struct RoundImpl<T, RoundMode::HALF_TO_ODD> {
+  static constexpr enable_if_floating_point<T> Round(const T val) {
+    return std::floor(val * T(0.5)) + std::ceil(val * T(0.5));
+  }
+};
+
+// Specializations of kernel state for round kernels
+template <typename>
+struct RoundOptionsWrapper;
+
+template <>
+struct RoundOptionsWrapper<RoundOptions> : public OptionsWrapper<RoundOptions> {
+  using OptionsType = RoundOptions;
+  double pow10;
+
+  explicit RoundOptionsWrapper(OptionsType options) : OptionsWrapper(std::move(options)) {
+    // Only positive of powers of 10 are used because combining multiply and
+    // division operations produced more stable rounding than using multiply-only.
+    // Refer to NumPy's round implementation:
+    // https://github.com/numpy/numpy/blob/7b2f20b406d27364c812f7a81a9c901afbd3600c/numpy/core/src/multiarray/calculation.c#L589
+    pow10 = RoundUtil::Pow10(std::abs(options.ndigits));
+  }
+
+  static Result<std::unique_ptr<KernelState>> Init(KernelContext* ctx,
+                                                   const KernelInitArgs& args) {
+    if (auto options = static_cast<const OptionsType*>(args.options)) {
+      return arrow::internal::make_unique<RoundOptionsWrapper<OptionsType>>(*options);
+    }
+
+    return Status::Invalid(
+        "Attempted to initialize KernelState from null FunctionOptions");
+  }
+};
+
+template <>
+struct RoundOptionsWrapper<RoundToMultipleOptions>
+    : public OptionsWrapper<RoundToMultipleOptions> {
+  using OptionsType = RoundToMultipleOptions;
+
+  static Result<std::unique_ptr<KernelState>> Init(KernelContext* ctx,
+                                                   const KernelInitArgs& args) {
+    ARROW_ASSIGN_OR_RAISE(auto state, OptionsWrapper<OptionsType>::Init(ctx, args));
+    auto options = Get(*state);
+    if (options.multiple <= 0) {
+      return Status::Invalid("Rounding multiple has to be a non-zero positive value");
+    }
+    return std::move(state);
+  }
+};
+
+template <RoundMode RndMode>
+struct Round {
+  using State = RoundOptionsWrapper<RoundOptions>;
+
+  template <typename T, typename Arg>
+  static enable_if_floating_point<Arg, T> Call(KernelContext* ctx, Arg arg, Status* st) {
+    static_assert(std::is_same<T, Arg>::value, "");
+    // Do not process Inf or NaN because they will trigger the overflow error at end of
+    // function.
+    if (!std::isfinite(arg)) {
+      return arg;
+    }
+    auto state = static_cast<State*>(ctx->state());
+    auto options = state->options;
+    auto pow10 = T(state->pow10);
+    auto round_val = (options.ndigits >= 0) ? (arg * pow10) : (arg / pow10);
+    // Use std::round() if in tie-breaking mode and scaled value is not 0.5.
+    if ((options.round_mode >= RoundMode::HALF_DOWN) &&
+        !RoundUtil::IsHalfInteger(round_val)) {
+      round_val = std::round(round_val);
+    } else if (!RoundUtil::IsInteger(round_val)) {
```

Comment:
I wonder: is this condition still useful? If `std::floor(round_val) == round_val`, then presumably the same also holds for `std::ceil` and `std::trunc`, so `RoundImpl::Round` will be a no-op.

Reply:
The integer logic is now fixed, so if the scaled value `round_val` is an integer then no rounding needs to take place. I also inlined the logic of `IsInteger` and `IsHalfInteger` so that the internal `floor` operation is done once.